include/stl2/detail/memory/destroy.hpp

```diff
@@ -0,0 +1,76 @@
+// cmcstl2 - A concept-enabled C++ standard library
+//
+//  Copyright Casey Carter 2015
+//  Copyright Christopher Di Bella 2016
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/caseycarter/cmcstl2
+//
+#ifndef STL2_DETAIL_ALGORITHM_DESTROY_HPP
+#define STL2_DETAIL_ALGORITHM_DESTROY_HPP
+#include <stl2/concepts.hpp>
+#include <stl2/iterator.hpp>
+#include <stl2/detail/fwd.hpp>
+#include <stl2/detail/memory/reference_to.hpp>
+
+STL2_OPEN_NAMESPACE {
+   ///////////////////////////////////////////////////////////////////////////
+   // destroy_at [Extension]
+   //
+   template <Destructible T>
+   void destroy_at(T* p) noexcept
+   {
+      p->~T();
+   }
+
+   ///////////////////////////////////////////////////////////////////////////
+   // destroy_at [Extension]
+   //
+   template <Destructible T>
+   void destroy_at(T& p) noexcept
+   {
+      p.~T();
+   }
+
+   ///////////////////////////////////////////////////////////////////////////
+   // destroy [Extension]
+   //
+   template <InputIterator I, Sentinel<I> S>
+   requires
+      Destructible<value_type_t<I>>() &&
+      __ReferenceTo<I, value_type_t<I>>()
+   I destroy(I first, S last) noexcept
+   {
+      for (; first != last; ++first)
+         destroy_at(__stl2::addressof(*first));
```

Comment:
ADL is the bane of my C++ programming: I know it exists, I generally know how it works, but can never remember the specifics. It's high time that I learn its rules properly.